### DISCUSSION TITLE
Indicator for detecting support and resistance

### DISCUSSION THREAD

**Discussion by PAKAMPHAHLELA** [kernc/backtesting.py#743] [Q&A]
has anybody been able to comeup with a code for detecting market key levels , then intergrate it within the backtesting.py framework for testing?

**Comment by pranavlal** [comment]
Hi,
There are various ways to detect support and resistance levels. I do have an experimental implementation but before I publish that here, I need one question answered. Is it possible to know where I am in the dataset? That is, when the next method is called, which is the row of the dataset for which the price is being output?
The support and resistances do not have one-to-one mapping to the prices. So, I need to find the closest price to the current row under consideration hence my question.

**Reply by kernc** [reply]
[`len(Strategy.data)`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.data&gsc.tab=0) is incremented on every `next()` iteration and indicates the number of bars up to now. Hope this helps.

**Comment by pranavlal** [comment]
Hi,

I tried the following code from within the next routine.

print("current iteration=",len(Strategy.data))                                                                 

and got the following error message.     

TypeError: object of type 'property' has no len()                                                                       

(

 

What am I doing wrong?

**Comment by pranavlal** [comment]
Hi all,
I figured out the problem with the code. I had to use
[code]
len(self.data)
[/code]
to get the length. 

Here is the code I have so far. I am not sure if my logic of calculating the support and or resistance price for the current closing price works.
The code does run.

[code]
import sys
from backtesting import Backtest, Strategy
from backtesting.lib import TrailingStrategy
from backtesting.test import GOOG
from identify_candlestick import recognize_candlestick
import recognize_candlestick_patterns 
import talib
import numpy as np
import pandas as pd
from datetime import datetime

dateparse = lambda x: datetime.strptime(x, '%Y-%m-%d %H:%M:%S') #To handle hourly data in pandas
def take_closest(myList, myNumber):
	"""
	Assumes myList is sorted. Returns closest value to myNumber.

	If two numbers are equally close, return the smallest number.
	"""
	pos = bisect_left(myList, myNumber)
	if pos == 0:
		return myList[0]
	if pos == len(myList):
		return myList[-1]
	before = myList[pos - 1]
	after = myList[pos]
	if after - myNumber < myNumber - before:
		return after
	else:
		return before

def isSupport(df,i):
	#routine taken from https://towardsdatascience.com/detection-of-price-support-and-resistance-levels-in-python-baedc44c34c9
	support = df['Low'][i] < df['Low'][i-1]  and df['Low'][i] < df['Low'][i+1] and df['Low'][i+1] < df['Low'][i+2] and df['Low'][i-1] < df['Low'][i-2]
	return support
def isResistance(df,i):
	#Routine taken from https://towardsdatascience.com/detection-of-price-support-and-resistance-levels-in-python-baedc44c34c9
	resistance = df['High'][i] > df['High'][i-1]  and df['High'][i] > df['High'][i+1] and df['High'][i+1] > df['High'][i+2] and df['High'][i-1] > df['High'][i-2]
	return resistance
def detect_support_resistances(df):

	support_indexes=[]
	resistance_indexes=[]
	support_prices=[]
	resistance_prices=[]
	for i in range(2,df.shape[0]-2):
	
		if isSupport(df,i):
			
			support_indexes.append(i)
			support_prices.append(df["Low"].iat[i])

		elif isResistance(df,i):

			resistance_indexes.append(i)
			resistance_prices.append(df["High"].iat[i])

	return support_indexes,support_prices,resistance_indexes,resistance_prices

class candlestickAdx(TrailingStrategy):

	atr_period=14
	rsi_lookback=14
	stop_loss_factor=2
	take_profit_factor=3
	l_ema_lookback=25
	s_ema_lookback=13
	signal_lookback=12
	bbands_period=20
	cci_period=14
	kc_lookback=14
	kc_multiplier=2
	stochastic_period=14
	def init(self):
		super().init()
		self.upper_band,self.middle_band,self.lower_band=self.I(talib.BBANDS,self.data.df["close"].to_numpy(),timeperiod=self.bbands_period)
		self.cci=self.I(talib.CCI,self.data.High,self.data.Low,self.data.Close,self.cci_period)
		self.support_index,self.support_price,self.resistance_index,self.resistance_price=detect_support_resistances(self.data.df)
		self.sto_d,self.sto_k=self.I(add_stochastic_oscillator,self.data.df,self.stochastic_period)
		

		self.atr=self.I(talib.ATR,self.data.High,self.data.Low,self.data.Close,timeperiod=self.atr_period)
		self.s_ma=self.I(talib.MA,self.data.Close,timeperiod=self.s_ema_lookback)
		self.kc_middle, self.kc_upper, self.kc_lower=self.I(get_kc,self.data.df["High"],self.data.df["Low"],self.data.df["Close"],self.kc_lookback,self.kc_multiplier,self.atr_period)
		self.rsi=self.I(talib.RSI,self.data.df["close"].to_numpy(),timeperiod=self.rsi_lookback)
		self.candlestick_pattern=self.data.df["candlestick_pattern"]		
		self.moddern_candlestick_patterns=recognize_candlestick_patterns.identify_moddern_candlestick_patterns(self.data.df,self.atr_period)
		
		self.set_trailing_sl(self.stop_loss_factor)
		#self.f_levels=self.I(fibonachi_retracements,self.data.df)
		self.tsi,self.tsi_signal=get_tsi(self.data.df["close"], self.l_ema_lookback, self.s_ema_lookback, self.signal_lookback)
		
		
	def next(self):
		super().next() 
		limit_order_target=(self.data.High-self.data.Low)+self.data.High
		stop_loss=self.data.Close-self.stop_loss_factor*self.atr
		take_profit=self.data.Close+self.take_profit_factor*self.atr
		
		current_support_price=take_closest(self.support_price,self.data.Close)
		current_resistance_price=take_closest(self.resistance_price,self.data.Close)
		#(("bull" in str(self.moddern_candlestick_patterns).lower())) and 
		if  self.data.Close>current_support_price:
			
			self.buy(sl=stop_loss,tp=take_profit)
		elif ((self.data.Close>=take_profit)  or (self.data.Close>current_resistance_price)):

			self.position.close()    
df=pd.read_csv("./ohlc/hourly/SBIN.csv",parse_dates=['date'], index_col=['date'],date_parser=dateparse)
print("length of data frame=",len(df))

#Run the candlestick recognition routines before back-testing
temp_frame=recognize_candlestick(df)
df["candlestick_pattern"]=temp_frame["candlestick_pattern"]
patterns_list=df["candlestick_pattern"].unique().tolist()
with open("patterns.txt", 'w') as fp:
	fp.writelines(patterns_list)
df["Open"]=df["open"]
df["High"]=df["high"]
df["Low"]=df["low"]
df["Close"]=df["close"]


bt = Backtest(df, candlestickAdx, commission=.002,exclusive_orders=True,cash=30000)

opt=bt.optimize(take_profit_factor=range(2,3),stop_loss_factor=range(1,3),atr_period=range(14,22),rsi_lookback=range(10,40),stochastic_period=range(14,30),maximize='Equity Final [$]',random_state=0,max_tries=300)
o=bt.run()
print(o)
#print(opt._strategy)
#bt.plot(filename="strategy.html")

[/code]
